runtime/compiler/env/j9method.cpp

```diff
@@ -4166,6 +4166,13 @@ void TR_ResolvedJ9Method::construct()
       {  TR::unknownMethod},
       };
 
+   static X JavaUtilRegexMatcherMethods [] =
+      {
+      {x(TR::java_util_regex_Matcher_init, "<init>", "(Ljava/util/regex/Pattern;Ljava/lang/CharSequence;)V")},
```

Comment:
This change does not appear to be safe for JDK11 which has a third array which is of type `IntHashSet[]`. In the `reset()` method we call the `clear()` API if the array index is non-null [2]. This means if we are no zero initializing the array, then it could have garbage data in there and we would try to call the `clear()` method on a garbage object. This should be easily reproducible in a test.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/c9bd6c8ae0bee2fb301c6af09cb7a23099f4cd16/src/java.base/share/classes/java/util/regex/Matcher.java#L250
[2] https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/c9bd6c8ae0bee2fb301c6af09cb7a23099f4cd16/src/java.base/share/classes/java/util/regex/Matcher.java#L410-L413

Reply:
More generally, how do ensure that future changes to JCL methods will not break us? What happens when for example in JDK22 a new array gets introduced in this type and the author expects it to be zero initialized?

This seems like something that the JIT should be deducing, at least for non-OpenJ9 JCL methods, otherwise we are just waiting for silent bugs to be introduced.

Reply:
It isn't possible to deduce this programattically when values are in fields and potentially accessed on multiple threads using a local anlaysis - whitelisting is the only option. There is a risk form future JCL change, but this method is not unlike many other core JCL methods that need special handling.

Thanks for pointing out the extra array in JDK11 - given that it is a reference array - those do not support zero init skipping -the option only affects primitives which is why the testing works and the optimization works in general. I believe there are also other methods relying on this behavior.

Reply:
> Thanks for pointing out the extra array in JDK11 - given that it is a reference array - those do not support zero init skipping -the option only affects primitives which is why the testing works and the optimization works in general.

Just to clarify explicitly, we only allow skip zero initialization on primitive arrays? Could you point to the code where this check happens for future reference? I did not know we had this restriction, but happy that we do.

Reply:
Dug it up:
https://github.com/eclipse/openj9/blob/c0ec632cad6b4408118467b2542e762efd7fa5e0/runtime/compiler/ilgen/Walker.cpp#L5980-L5981

This is where we check if the method symbol has been marked for the JIT to allow skipping zero initialization. It is used when new (primitive) arrays are created during ILGen. The above concern is no longer valid.